// @luxonis/common-fe-utils/src/utils/apps.ts
export function hasLocalFrontend(appIdentifier: string): boolean {
  return isDefaultApp(appIdentifier) || isViewerApp(appIdentifier);
}

export function isDefaultApp(appIdentifier: string): boolean {
  return appIdentifier === 'com.luxonis.default' || appIdentifier === 'com.stg.luxonis.default';
}

export function isViewerApp(appIdentifier: string): boolean {
  return appIdentifier === 'com.luxonis.viewer' || appIdentifier === 'com.stg.luxonis.viewer';
}

export function isStageApplication(appIdentifier: string): boolean {
  return appIdentifier.startsWith('com.stg.');
}

export function constructLocalFrontendUrl(identifier: string, token: string, clientId: string, version: string) {
  const prefilledUrlTail = `${version}/?t=${token}&cid=${clientId}`;

  if (isDefaultApp(identifier)) {
    const [major, minor, _patch] = version.split('.').map(Number);

    return (major === 0 && minor >= 9) || major > 0
      ? `https://public.luxonis.app/viewer/${prefilledUrlTail}`
      : `https://public.luxonis.app/visualizer/${prefilledUrlTail}`;
  }

  if (isViewerApp(identifier)) {
    return `https://public.luxonis.app/viewer/${prefilledUrlTail}`;
  }

  return undefined;
}

const APP_STARTED_EVENT = 'app-started-' as const;
export function createAppStartedEventEmitterMessage(appId: string) {
  return `${APP_STARTED_EVENT}${appId}`;
}
