// @luxonis/common-fe-utils/src/utils/apps.ts
export function hasLocalFrontend(appIdentifier: string | undefined): boolean {
  if (!appIdentifier) {
    return false;
  }

  return isDefaultApp(appIdentifier) || isViewerApp(appIdentifier);
}

export function isDefaultApp(appIdentifier: string): boolean {
  return appIdentifier === 'com.luxonis.default' || appIdentifier === 'com.stg.luxonis.default';
}

export function isViewerApp(appIdentifier: string): boolean {
  return appIdentifier === 'com.luxonis.viewer' || appIdentifier === 'com.stg.luxonis.viewer';
}

export function isStageApplication(appIdentifier: string): boolean {
  return appIdentifier.startsWith('com.stg.');
}

export function constructLocalFrontendUrl(
  identifier: string,
  token: string,
  clientId: string,
  version: string,
  env: 'prod' | 'stg',
  domain?: string,
) {
  const prefilledUrlTail = `${version}/?t=${token}&cid=${clientId}`;

  if (domain) {
    return `${domain}${prefilledUrlTail}`;
  }

  if (isDefaultApp(identifier)) {
    return env === 'prod'
      ? `https://viewer.luxonis.app/${prefilledUrlTail}`
      : `https://viewer.stg.luxonis.app/${prefilledUrlTail}`;
  }

  return undefined;
}

const APP_STARTED_EVENT = 'app-started-' as const;
export function createAppStartedEventEmitterMessage(appId: string) {
  return `${APP_STARTED_EVENT}${appId}`;
}

export function formatAppName(identifier: string, name?: string | null | undefined) {
  return name ?? identifier;
}
