// @luxonis/common-fe-utils/src/utils/validation.ts
import { z } from 'zod';

export type ValidationError =
  | {
      type: 'invalid_length';
      atLeast?: number;
      atMost?: number;
    }
  | {
      type: 'missing';
    }
  | {
      type: 'invalid_value';
      expected: 'email' | 'redirect_path';
      received: string;
    };

export type ValidationErrorType = ValidationError['type'];

export type ValidateResult<T> = (ValidationError & { is: 'error' }) | { is: 'valid'; value: T };

export const zodPassword = z.string().min(6);

export const zodSemVerFormat = z
  .string()
  .regex(/^(0|[1-9]\d*)\.(0|[1-9]\d*)(?:\.(0|[1-9]\d*))?$/, { message: 'Version must be in the format X.Y or X.Y.Z' });

const REDIRECT_URL_PARAM_REGEX = new RegExp(/^\/[a-zA-Z0-9/-]*\??([a-zA-Z0-9]+=[a-zA-Z0-9]+)*$/);

export function validateRedirectUrlParam(value: unknown): ValidateResult<string> {
  if (typeof value !== 'string') {
    return { is: 'error', type: 'missing' };
  }

  if (!REDIRECT_URL_PARAM_REGEX.test(value)) {
    return { is: 'error', type: 'invalid_value', expected: 'redirect_path', received: value };
  }

  return { is: 'valid', value };
}
